### PR TITLE
Feature/update sim model

### PR DESCRIPTION
**Pull request by HansRobo**
## Types of PR

- [ ] New Features
- [x] Upgrade of existing features
- [ ] Bugfix

## Description

Import `simple_planning_simulator` changes except https://github.com/autowarefoundation/autoware.universe/pull/4941

## Others

To use `DELAY_STEER_MAP_ACC_GEARED` model, `acceleration_map_path` should be set in simulator parameter file of description package like below.
https://github.com/autowarefoundation/autoware.universe/blob/main/simulator/simple_planning_simulator/param/simple_planning_simulator_default.param.yaml#L20

### CODE REVIEW

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+#include <queue>
+#include <simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_interface.hpp>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "eigen3/Eigen/Core"
+#include "eigen3/Eigen/LU"
+
+// copied from https://github.com/tier4/autoware.universe/blob/v0.17.0/common/interpolation/include/interpolation/interpolation_utils.hpp
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp:31]: Please use doxygen [note command.](https://www.doxygen.nl/manual/commands.html#cmdnote)

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+  // when query_keys is out of base_keys (This function does not allow exterior division.)
+  constexpr double epsilon = 1e-3;
+  if (
+    query_keys.front() < base_keys.front() - epsilon ||
+    base_keys.back() + epsilon < query_keys.back()) {
+    throw std::invalid_argument("query_keys is out of base_keys");
+  }
+
+  // NOTE: Due to calculation error of double, a query key may be slightly out of base keys.
+  //       Therefore, query keys are cropped here.
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp:92]: Please use doxygen [note command.](https://www.doxygen.nl/manual/commands.html#cmdnote)

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+              << csv_path << std::endl;
+    return true;
+  }
+
+  double getAcceleration(const double acc_des, const double vel) const
+  {
+    std::vector<double> interpolated_acc_vec;
+    const double clamped_vel = CSVLoader::clampValue(vel, vel_index_, "acc: vel");
+
+    // (throttle, vel, acc) map => (throttle, acc) map by fixing vel
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp:187]: Please use doxygen [note command.](https://www.doxygen.nl/manual/commands.html#cmdnote)

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+     */
+  SimModelDelaySteerMapAccGeared(
+    double vx_lim, double steer_lim, double vx_rate_lim, double steer_rate_lim, double wheelbase,
+    double dt, double acc_delay, double acc_time_constant, double steer_delay,
+    double steer_time_constant, std::string path);
+
+  /**
+     * @brief default destructor
+     */
+  ~SimModelDelaySteerMapAccGeared() = default;
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp:232]: Please fix indent.

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+     * @brief update state considering current gear
+     * @param [in] state current state
+     * @param [in] prev_state previous state
+     * @param [in] gear current gear (defined in autoware_auto_msgs/GearCommand)
+     * @param [in] dt delta time to update state
+     */
+  void updateStateWithGear(
+    Eigen::VectorXd & state, const Eigen::VectorXd & prev_state, const uint8_t gear,
+    const double dt);
+};
```
Comment (hakuturu583) [inline simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp:336]:
```suggestion
    /**
     * @brief set queue buffer for input command
     * @param [in] dt delta time
     */
  void initializeInputQueue(const double & dt);

    /**
     * @brief get vehicle position x
     */
  double getX() override;

    /**
     * @brief get vehicle position y
     */
  double getY() override;

  /**
     * @brief get vehicle angle yaw
     */
  double getYaw() override;

    /**
     * @brief get vehicle velocity vx
     */
  double getVx() override;

    /**
     * @brief get vehicle lateral velocity
     */
  double getVy() override;

    /**
     * @brief get vehicle longitudinal acceleration
     */
  double getAx() override;

    /**
     * @brief get vehicle angular-velocity wz
     */
  double getWz() override;

    /**
     * @brief get vehicle steering angle
     */
  double getSteer() override;

    /**
     * @brief update vehicle states
     * @param [in] dt delta time [s]
     */
  void update(const double & dt) override;

    /**
     * @brief calculate derivative of states with time delay steering model
     * @param [in] state current model state
     * @param [in] input input vector to model
     */
  Eigen::VectorXd calcModel(const Eigen::VectorXd & state, const Eigen::VectorXd & input) override;

    /**
     * @brief update state considering current gear
     * @param [in] state current state
     * @param [in] prev_state previous state
     * @param [in] gear current gear (defined in autoware_auto_msgs/GearCommand)
     * @param [in] dt delta time to update state
     */
  void updateStateWithGear(
    Eigen::VectorXd & state, const Eigen::VectorXd & prev_state, const uint8_t gear,
    const double dt);
};
```

Comment (hakuturu583) [reply simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp:336]: Please fix indent.

**Review by hakuturu583 — CHANGES_REQUESTED** [review]

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp (hunk 2 above)
Comment (HansRobo) [reply simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp:92]: This is imported from `simple_planning_simulator` original file, so it is not desirable to fix.

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp (hunk 3 above)
Comment (HansRobo) [reply simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp:187]: This is imported from simple_planning_simulator original file, so it is not desirable to fix.

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp (hunk 1 above)
Comment (HansRobo) [reply simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp]: fix in https://github.com/tier4/scenario_simulator_v2/pull/1144/commits/8f992dc0fd474ac599a9470102e5abc78d42310f

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp (hunk 4 above)
Comment (HansRobo) [reply simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp]: fixed in https://github.com/tier4/scenario_simulator_v2/pull/1144/commits/1d8fdcf70587c5636d2b327722005e15ebbc42d4

File: simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp (hunk 3 above)
Comment (HansRobo) [reply simulation/simple_sensor_simulator/include/simple_sensor_simulator/vehicle_simulation/vehicle_model/sim_model_delay_steer_map_acc_geared.hpp:192]: In this place, I found that the note command is invalid to `doxygen`.

**Review by hakuturu583 — APPROVED** [review]

**Pull request merged**